Add f(x) = A*x, A\x

alyst opened the pull request:
The PR adds `f(x) = A*x` and `f(x) = A\x` functions, where `A` is a constant matrix, and `x` is either a vector or a matrix.
For vector arguments, the new `VECTOR_TO_VECTOR_FUNCS` list of test functions is introduced.

The tested `A` matrices are: diagonal, dense, lower and upper triangular. The last three matrix types are represented by the [Lehmer matrix](https://en.wikipedia.org/wiki/Lehmer_matrix).

In particular, the `A\x` functions are useful to test the autodiff of multivariate normal distribution.
This PR would be required to properly test JuliaDiff/ForwardDiff.jl#589 (so when/if merged, the version number of *DiffTools.jl* has to be updated).

alyst committed:
tests: refine type checks

alyst committed:
add A*x to VEC_TO_VEC funcs

alyst committed:
add A*X to MAT_TO_MAT funcs

alyst commented:
@oxinabox @fredrikekre Could some of you please review this PR, so we can move forward with JuliaDiff/ForwardDiff.jl#589? Thank you!

alyst committed:
fix misplaced comment

alyst commented:
@mateuszbaran @mohamed82008 Could you pls take a look?

mateuszbaran commented on src/DiffTests.jl:
```diff
@@ -249,6 +249,39 @@ end
+    Diagonal(LinRange(convert(T, 0.01), convert(T, 100.0), n))
+diag_matrix(x::VecOrMat) = diag_matrix(Float64, size(x, 1))
+
+hilbert_matrix(::Type{T}, n::Integer) where T<:Real =
+    [convert(T, inv(i + j - 1)) for i in 1:n, j in 1:n]
+hilbert_matrix(x::VecOrMat) = hilbert_matrix(Float64, size(x, 1))
```
This function doesn't seem to be used anywhere in this PR.

alyst replied:
> This function doesn't seem to be used anywhere in this PR.

Yes, it doesn't. I was undecided which constant matrix to use, it looked like with this one the errors are higher (it is beyond my capability to track and possibly fix the source of that errors).
I can remove, if it is critical, but maybe it is better to keep it, anticipating the future extension of the test coverage.

mateuszbaran replied:
I generally think keeping unused code in the main branch should be avoided. Everything else looks fine though in case someone with merge rights cares about my opinion :slightly_smiling_face: .

oxinabox replied:
I agree, this should be removed before we can merge

alyst replied:
@oxinabox done

alyst commented:
@oxinabox @fredrikekre gentle ping to anyone with PR merge rights :) Please help me to reduce the stack of dependent PRs that I have accumulated :)

alyst committed:
add sparse matrix-based functions

alyst committed:
remove hilbert_matrix() (unused)

alyst commented:
@oxinabox Gentle ping :) I have removed the unused test function; also I have added the sparse variants of the new test functions.
Could you please merge if there are no further adjustments to make?

alyst commented:
@oxinabox Thanks! Can you tag a new release?
